File: 08-11-2025/2Dbfs.cpp
// #include <bits/stdc++.h>
// using namespace std;

// /************************************************************
// *                                                          *
// *  "If talent doesn't work, believe in yourself and        *
// *   do hard work. Allah will give you the best gift."      *
// *                                                          *
// *************************************************************/
// #define ll long long
// #define hea cout << "YES\n";
// #define na cout << "NO\n";
// #define nl cout << '\n';

// const int N = 1005;
// char grid[N][N];
// bool vis[N][N];
// int level[N][N];

// vector<pair<int, int>> dir = {{0, 1}, {0, -1}, {-1, 0}, {1, 0}};
// int n, m;

// bool vaild(int x, int y)
// {
//     return(x >= 0 && x < n && y >= 0 && y < m);
// }

// void bfs(int si, int sj)
// {
//     queue<pair<int, int>> q;
//     q.push({si, sj});
//     vis[si][sj] = true;
//     level[si][sj] = 0;
    
//     while(!q.empty())
//     {
//         auto par = q.front();
//         q.pop();

//         // cout << par.first << " " << par.second << '\n';
//         int x = par.first, y = par.second;
//         for(int i = 0; i < 4; i++)
//         {
//             int par_i = x + dir[i].first, par_j = y + dir[i].second;
//             if(vaild(par_i, par_j) && !vis[par_i][par_j])
//             {
//                 q.push({par_i, par_j});
//                 vis[par_i][par_j] = true;
//                 level[par_i][par_j] = level[x][y] + 1;
//             }
//         }
//     }
// }

// int main() 
// {
//     ios::sync_with_stdio(0), cin.tie(0);

//     // sieve(); 
//     cin >> n >> m;
//     for(int i = 0; i < n; i++)
//     {
//         for(int j = 0; j < m; j++)
//         {
//             cin >> grid[i][j];
//         }
//     }

//     memset(vis, false, sizeof(vis));
//     int si, sj;
//     cin >> si >> sj;

//     bfs(si, sj);
//     int di, dj;
//     cin >> di >> dj;

//     cout << level[di][dj] << '\n';

//     return 0;
// }


#include <bits/stdc++.h>
using namespace std;

/************************************************************
*                                                          *
*  "If talent doesn't work, believe in yourself and        *
*   do hard work. Allah will give you the best gift."      *
*                                                          *
*************************************************************/
#define ll long long
#define hea cout << "YES\n";
#define na cout << "NO\n";
#define nl cout << '\n';

int n, m;
const int N = 1005;
char grid[N][N];
bool vis[N][N];
int parent[N][N];

vector<pair<int, int>> dir = {{0, 1}, {0, -1}, {-1, 0}, {1, 0}};

bool valid(int x, int y)
{
    return(x >= 0 && x < n && y >= 0 && y < m);
}

void bfs(int si, int sj)
{
    queue<pair<int, int>> q;
    q.push({si, sj});
    vis[si][sj] = true;
    parent[si][sj] = 0;

    while(!q.empty())
    {
        auto p = q.front();
        q.pop();

        int par_i = p.first, par_j = p.second;
        for(int i = 0; i < 4; i++)
        {
            int child_i = par_i + dir[i].first;
            int child_j = par_j + dir[i].second;

            if(valid(child_i, child_j) && !vis[child_i][child_j])
            {
                cout << child_i << " " << child_j << "\n";
                q.push({child_i, child_j});
                vis[child_i][child_j] = true;
                parent[child_i][child_j] = parent[par_i][par_j] + 1;
            }
        }
    }
}

int main() 
{
    ios::sync_with_stdio(0), cin.tie(0);

    // sieve(); 
    cin >> n >> m;
    for(int i = 0; i < n; i++)
    {
        for(int j = 0; j < m; j++)
        {
            cin >> grid[i][j];
        }
    }
    memset(vis, false, sizeof(vis));
    memset(parent, -1, sizeof(parent));

    int si, sj;
    cin >> si >> sj;

    bfs(si, sj);
    int di, dj;
    cin >> di >> dj;


    return 0;
}

File: 08-11-2025/2Dbfs.py
from collections import deque

# ------------------------------------------------------------
#  "If talent doesn't work, believe in yourself and
#   do hard work. Allah will give you the best gift."
# ------------------------------------------------------------

# Global variables
N = 1005
grid = [['' for _ in range(N)] for _ in range(N)]
vis = [[False for _ in range(N)] for _ in range(N)]
parent = [[-1 for _ in range(N)] for _ in range(N)]

# Directions (right, left, up, down)
dir = [(0, 1), (0, -1), (-1, 0), (1, 0)]

def valid(x, y, n, m):
    return 0 <= x < n and 0 <= y < m

def bfs(si, sj, n, m):
    q = deque()
    q.append((si, sj))
    vis[si][sj] = True
    parent[si][sj] = 0

    while q:
        par_i, par_j = q.popleft()
        for dx, dy in dir:
            child_i = par_i + dx
            child_j = par_j + dy

            if valid(child_i, child_j, n, m) and not vis[child_i][child_j]:
                print(child_i, child_j)  # same as C++ cout
                q.append((child_i, child_j))
                vis[child_i][child_j] = True
                parent[child_i][child_j] = parent[par_i][par_j] + 1


# ------------------ main ------------------
if __name__ == "__main__":
    n, m = map(int, input().split())
    for i in range(n):
        row = list(input().strip())
        for j in range(m):
            grid[i][j] = row[j]

    # reset visited and parent arrays
    for i in range(n):
        for j in range(m):
            vis[i][j] = False
            parent[i][j] = -1

    si, sj = map(int, input().split())
    bfs(si, sj, n, m)

    di, dj = map(int, input().split())
